### PR TITLE
feat(logger): expose FilterConfig injection at app bootstrap

### LINKED ISSUE #430: log: Allow consumer to inject logger.FilterConfig at app bootstrap (PCI masking blocker)
## Summary

The framework constructs the application logger via a hardcoded `logger.New(level, pretty)` call inside `app.Builder.CreateLogger()`, with no public surface to inject a `*logger.FilterConfig`. The framework already ships `logger.NewWithFilter(level, pretty, *FilterConfig)` and a fully-featured `SensitiveDataFilter`, but consumers cannot reach it without either forking, monkey-patching, or wrapping `deps.Logger` per-module — and none of those cover the framework's own emit paths (HTTP middleware, AMQP consumer wrappers, panic recovery, etc.).

This issue proposes a small, additive surface to let applications configure the logger filter at bootstrap, so the existing `SensitiveDataFilter` plumbing actually applies to every log line — not just lines emitted from code the consumer wrote.

## Background — who I am and what I'm building

I'm building [`internal-tokenization-itsp-api-go`](https://jira4novo.atlassian.net/browse/NGT-4943): a Go consolidation of ~9 legacy Java microservices into a single ITSP (Issuer Token Service Provider) for Visa VTS. It's built on `go-bricks v0.30.0` and follows the demo project's layout exactly.

The service handles primary account numbers (PAN), card verification values (CVV2), one-time passwords (OTP), tokens, and other PCI-sensitive payloads. Compliance constraints make automatic, framework-level masking a hard requirement rather than a nice-to-have.

## Use case — why this matters

**Acceptance criterion from the upstream JIRA story (NGT-4943, Foundation):**

> Logging estructurado (Zerolog vía go-bricks) con **máscara automática de PAN/CVV/OTP**.

Two PCI-DSS v4.0 requirements drive the "automatic" qualifier:

- **3.4.1 / 3.5.1** — PAN must be rendered unreadable anywhere it is stored, including audit logs. Full PAN may only appear with documented business justification.
- **3.3.1** — Sensitive authentication data (CVV2, full track, PIN, OTP) must **never** be retained post-authorization, including in logs.

"Automatic" here means: a developer who emits `log.Str("pan", req.PAN)` in a handler — or who accidentally embeds a PAN inside an `errors.New("card %s failed", req.PAN)` that gets logged — must produce a masked output without code review catching every callsite. Field-name-based masking via `SensitiveDataFilter` is the standard tool for this, and the framework already implements it. The gap is purely in how a consumer hands a `*FilterConfig` to the framework.

The risk profile if this stays manual:
- Every PR touching log statements becomes a PCI review surface.
- The framework's own middleware (request/response logging in `server.LoggerWithConfig`, panic recovery in AMQP, slow-request warnings) emits log lines I cannot intercept by wrapping `deps.Logger` after the fact — those subsystems captured the logger pointer at startup.
- Any future framework subsystem that logs (cache manager, scheduler, outbox) inherits the unfiltered logger.

## Current state of the framework

Tracing the logger construction path in `v0.30.0`:

```go
// app/app_builder.go:46-64
func (b *Builder) CreateLogger() *Builder {
    if b.err != nil { return b }
    if b.cfg == nil {
        b.err = fmt.Errorf("configuration required before creating logger")
        return b
    }
    b.logger = logger.New(b.cfg.Log.Level, b.cfg.Log.Pretty)  // ← hardcoded; no filter
    b.logger.Info().
        Str("app", b.cfg.App.Name).
        Str("env", b.cfg.App.Env).
        Str("version", b.cfg.App.Version).
        Msg("Starting application")
    return b
}
```

```go
// logger/logger.go:33, 70
func New(level string, pretty bool) *ZeroLogger { ... }
func NewWithFilter(level string, pretty bool, filterConfig *FilterConfig) *ZeroLogger { ... }
```

```go
// app/options.go:11-23 — no Logger or FilterConfig field
type Options struct {
    Database               database.Interface
    MessagingClient        messaging.Client
    SignalHandler          SignalHandler
    TimeoutProvider        TimeoutProvider
    Server                 ServerRunner
    ConfigLoader           func() (*config.Config, error)
    DatabaseConnector      func(*config.DatabaseConfig, logger.Logger) (database.Interface, error)
    MessagingClientFactory func(string, logger.Logger) messaging.AMQPClient
    CacheConnector         cache.Connector
    ResourceSource         TenantStore
}
```

The filter infrastructure is fully built (`logger/filter.go`: `FilterConfig`, `SensitiveDataFilter`, `DefaultFilterConfig`, case-insensitive substring matching, URL credential masking, cycle-safe recursive struct/map filtering) — it's just not reachable from `app.New()` / `app.NewWithOptions()` / `app.NewWithConfig()`.

## Current workarounds and why each is inadequate

I evaluated four approaches before opening this issue.

### 1. Wrap `deps.Logger` inside each module's `Init()`

Conceptually:

```go
func (m *Module) Init(deps *app.ModuleDeps) error {
    m.logger = sharedkernel.WrapWithFilter(deps.Logger, sensitive.LogFilterConfig())
    // ...
}
```

**Why it doesn't work for our use case:**
- Requires implementing a wrapper for `logger.Logger` + `LogEvent` (~10 methods on each interface). Significant maintenance surface against an evolving interface.
- Wraps **only** the logger used by the module's own code. The framework retains a separate pointer to the unwrapped logger for: `server.LoggerWithConfig` middleware (HTTP request logs), `messaging` consumer panic recovery, slow-request warnings, the bootstrap `Starting application` line, and every internal `b.logger.Info()`/`Warn()` call in the app package.
- Module load order matters: if the wrap module loads after a module that consumed `deps.Logger`, the earlier module gets the unwrapped logger.
- PCI-DSS audits typically request a single point of evidence that masking is applied at the I/O boundary — "every module remembers to wrap" is not auditable.

### 2. Mutate `deps.Logger` inside an early-registered "shared-kernel" module

Same problem as (1): the framework's internal subsystems captured the logger reference during `bootstrap.dependencies()` and `Builder.CreateLogger()` and won't pick up mutations.

### 3. Stop using `app.New()` and run my own bootstrap

I could replicate the relevant bits of `Builder` in user code and call `logger.NewWithFilter` myself. But:
- `Builder` orchestrates ~12 dependencies (OTel provider, observability bridge, multitenant resolver, etc.) — almost all the framework value is in that orchestration.
- Several types (`appBootstrap`, `dependencyBundle`, `factoryResolver`) are unexported.
- This effectively forks the framework's bootstrap for every consumer that needs masking, defeating the point of go-bricks.

### 4. Manual masking at every call site

```go
log.Str("pan_masked", masking.PAN(req.PAN))
```

This is what I shipped in the first PR ([`internal/sharedkernel/sensitive`](https://jira4novo.atlassian.net/browse/NGT-4943) — `PAN`, `CVV`, `OTP`, `Scrub` helpers). It's defensive depth, not the AC requirement. It also doesn't help when:
- A third-party library logs through the framework's logger.
- An error message gets serialized into a log via `log.Err(err)` where `err.Error()` happens to contain a PAN.
- A handler logs a struct via `log.Interface("payload", req)` where the struct has a PAN-bearing field.

`SensitiveDataFilter` already handles all three cases. We just can't reach it.

## Proposed API — three options, pick whichever fits the framework's style

All three are additive and backwards-compatible.

### Option A — `Options.LoggerFilterConfig`

Smallest change. Add one field to `app.Options`:

```go
type Options struct {
    // ... existing fields ...

    // LoggerFilterConfig customizes the sensitive-data filter applied to the
    // framework logger. When nil, logger.DefaultFilterConfig() is used (current
    // behavior). To opt out entirely, pass &logger.FilterConfig{SensitiveFields: nil}.
    LoggerFilterConfig *logger.FilterConfig
}
```

Then in `Builder.CreateLogger`:

```go
if b.opts != nil && b.opts.LoggerFilterConfig != nil {
    b.logger = logger.NewWithFilter(b.cfg.Log.Level, b.cfg.Log.Pretty, b.opts.LoggerFilterConfig)
} else {
    b.logger = logger.New(b.cfg.Log.Level, b.cfg.Log.Pretty)
}
```

**Pros:** trivial diff, mirrors how `Database`, `Server`, etc. are injected via `Options`.
**Cons:** the consumer can't add fields to the framework's default list without restating them — though I'd argue exposing `logger.DefaultFilterConfig()` (already exported) and letting consumers `append(cfg.SensitiveFields, customFields...)` covers that idiom cleanly.

### Option B — `Options.Logger`

Let the consumer pre-construct the logger:

```go
type Options struct {
    // ... existing fields ...

    // Logger overrides the framework-constructed logger. When set, the
    // bootstrap skips Builder.CreateLogger entirely and uses this instance.
    Logger logger.Logger
}
```

**Pros:** maximally flexible — consumers can wire OTel, severity hooks, custom writers, etc. all together.
**Cons:** larger blast radius. The framework currently emits a `Starting application` line as a side effect of `CreateLogger`; that lifecycle invariant has to move. Also the OTel enhancement in `bootstrap.enhanceLoggerWithOTel` does a type assertion to `*ZeroLogger` — if a consumer hands in a wrapper, OTel injection silently no-ops.

### Option C — config-driven via koanf

```yaml
# config.yaml
logger:
  level: info
  pretty: false
  sensitive_fields:           # NEW
    - pan
    - primary_account_number
    - cvv
```

Wire `config.LogConfig` to grow a `SensitiveFields []string` field, and have `CreateLogger` build a `FilterConfig` from it (additive to `DefaultFilterConfig()`).

**Pros:** no code change in consumer apps — set the env vars/yaml and you're done. Aligns with the framework's existing pattern of preferring configuration over options for declarative behavior.
**Cons:** can't compose values that need code (e.g., a regex-based pattern matcher, or a hook). For our use case, string field names are sufficient — so C alone would unblock us.

**My preference:** **A or C** — both are smaller than B and either is sufficient. C has the lowest cognitive cost for consumers but slightly more framework surface. A is the most surgical. If implementing both isn't disruptive, the combination is ideal: config-driven by default, code-overridable when needed.

## Acceptance criteria (suggested)

- [ ] A consumer can construct an app such that **every** log line emitted by the framework or any registered module passes through a `SensitiveDataFilter` configured with consumer-supplied field names, **without** wrapping `deps.Logger`.
- [ ] When no consumer config is provided, behavior is unchanged from `v0.30.0` (no regression on `DefaultFilterConfig`).
- [ ] Documentation in `llms.txt` / README covers the new surface with a payment-processing example.
- [ ] Unit test: wire a custom field, emit `Str("pan", "4111...")` through `deps.Logger`, assert the field is masked.
- [ ] Integration test: same field via `server.LoggerWithConfig` request log.

## Out of scope (for this issue)

- **Content-pattern masking** (scanning string values for PAN-shaped Luhn-valid digit sequences). I implemented this on the consumer side as `sensitive.Scrub`; it's value-side and doesn't need framework support.
- **Per-tenant masking policies.** Useful eventually for multi-tenant compliance regimes, but orthogonal to the bootstrap-level wiring this issue covers.
- **Structured PCI-DSS report generation** from the catalog. Out of scope for the framework.

## References

- Consumer-side errors catalog + masking helpers I just shipped (uses the field-name vocabulary that needs to feed back into the framework): `internal/sharedkernel/errors`, `internal/sharedkernel/sensitive`.
- Existing framework primitives that this issue would expose: `logger.NewWithFilter`, `logger.FilterConfig`, `logger.SensitiveDataFilter`, `logger.DefaultFilterConfig` — all in `logger/filter.go` and `logger/logger.go`.
- Code path that needs the hook: `app/app_builder.go:46-64` (`Builder.CreateLogger`).
- PCI-DSS v4.0 (publicly available at pcisecuritystandards.org) — sections 3.3, 3.4, 3.5.

Happy to send a PR for whichever option you prefer. Option A is a ~10-line diff; Option C is closer to ~30 lines once config validation is included.

### PR DESCRIPTION
**Pull request by gaborage**
## Summary

- Adds `config.LogConfig.SensitiveFields []string` (YAML/env, additive to `logger.DefaultFilterConfig`) and `app.Options.LoggerFilterConfig *logger.FilterConfig` (code-level, full replacement) so consumers can extend the framework-level log masking without forking the bootstrap.
- `Builder.CreateLogger` now routes through `logger.NewWithFilter`; when neither seam is set the behavior is byte-for-byte equivalent to v0.30.0 (`NewWithFilter(nil)` falls back to `DefaultFilterConfig`).
- Precedence: `Options.LoggerFilterConfig` > `cfg.Log.SensitiveFields` > `DefaultFilterConfig`. Options is full replacement; YAML is additive merge.

Closes #430

## Why

Issue #430 motivates this from a PCI-DSS 3.3/3.4/3.5 angle — a service handling PAN/CVV2/OTP needs **every** log line (including framework middleware, AMQP panic recovery, slow-request warnings) to pass through the existing `SensitiveDataFilter`, not just the lines a module emits via its wrapped `deps.Logger`. The filter infrastructure already shipped; only the bootstrap-level injection seam was missing.

## What changed

**Code (commit 1):**
- `config/types.go` — `LogConfig.SensitiveFields []string` with koanf tags.
- `app/options.go` — `Options.LoggerFilterConfig *logger.FilterConfig`.
- `app/app_builder.go` — `resolveLoggerFilterConfig(opts, &cfg.Log)` helper; `CreateLogger` switches from `logger.New` to `logger.NewWithFilter`.
- `app/app_builder_test.go` — unit tests for the helper (defaults, both seams independently, precedence, opt-out, edge cases) plus smoke tests through `CreateLogger`.
- `llms.txt` — new \"Sensitive Data Filtering\" subsection under structured logging.

**Docs (commit 2):**
- `wiki/observability.md` — comprehensive end-user deep dive (default field list, two-seam injection, matching semantics, defense-in-depth, migration notes).
- `CLAUDE.md` — Security Guidelines pointer to the wiki section.
- `config.example.yaml` — `log.sensitive_fields` example with common PCI fields.

## Test plan

- [x] \`make check\` — lint + race + unit tests across all framework packages pass.
- [x] \`make check-all\` — framework + tool checks pass.
- [x] \`golangci-lint run --enable gosec ./app/... ./config/... ./logger/...\` — 0 issues.
- [x] \`govulncheck ./...\` — no vulnerabilities.
- [x] /simplify review (3 parallel agents: reuse, quality, efficiency) — 3 trivial fixes applied, no structural issues.
- [x] /security-audit — no new SQL, no new HTTP boundary, no secret material, no shutdown logic; clean across all categories.
- [x] YAML parses correctly via koanf (the framework's actual parser).
- [x] All seven \`TestResolveLoggerFilterConfig\` sub-tests pass (defaults, nil safety, options precedence, opt-out, additive merge, empty-equals-nil, partial-options-no-filter).
- [ ] Manual: verify a service that sets \`log.sensitive_fields: [pan]\` masks \`log.Str(\"pan\", \"4111…\").Msg(...)\` to \`***\` in stdout JSON.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Adds sensitive-data filtering for framework logs with configurable field masking for PCI/PII protection
  * Configuration via YAML (additive list) or runtime options (full replacement); runtime option takes precedence
  * Case-insensitive field-name substring matching across nested objects and URL user/password components
  * Example config includes common PCI/PII field names to mask

* **Documentation**
  * Observability and security guidance updated with masking behavior, configuration seams, and migration notes

[Review Change Stack](https://app.coderabbit.ai/change-stack/gaborage/go-bricks/pull/431)

### CODE REVIEW

File: app/app_builder.go [hunk 1]
```diff
@@ -69,6 +69,26 @@ func (b *Builder) CreateLogger() *Builder {
...
+//  3. nil — logger.NewWithFilter falls back to logger.DefaultFilterConfig
+//     (identical to the legacy logger.New path; preserved for back-compat).
+func resolveLoggerFilterConfig(opts *Options, cfg *config.LogConfig) *logger.FilterConfig {
+	if opts != nil && opts.LoggerFilterConfig != nil {
+		return opts.LoggerFilterConfig
+	}
+	if cfg != nil && len(cfg.SensitiveFields) > 0 {
+		base := logger.DefaultFilterConfig()
+		base.SensitiveFields = append(base.SensitiveFields, cfg.SensitiveFields...)
+		return base
```
[the remark this answers is not in the document]
Comment (grosales-np) [reply app/app_builder.go:107]: Addressed in 8e53b2a. Applied trim + drop-empty + case-insensitive dedup (both intra-config and against DefaultFilterConfig). Added 4 sub-tests covering each normalization branch. Confirmed the empty-string vector would have produced a silent total-mask outcome — agreed, this is a real footgun worth catching at the config-merge layer rather than relying on operators to write clean YAML.
